src/axum_websocket.rs

```diff
@@ -0,0 +1,883 @@
+            self.protocol = protocols
+                .into_iter()
+                // FIXME: This will often allocate a new `String` and so is less efficient than it
+                // could be. But that can't be fixed without breaking changes to the public API.
+                .map(Into::into)
+                .find(|protocol| {
+                    req_protocols
+                        .split(',')
+                        .any(|req_protocol| req_protocol.trim() == protocol)
+                })
+                .map(|protocol| match protocol {
+                    Cow::Owned(s) => HeaderValue::from_str(&s).unwrap(),
+                    Cow::Borrowed(s) => HeaderValue::from_static(s),
+                });
+        }
+
+        self
+    }
+
+    /// Provide a callback to call if upgrading the connection fails.
+    ///
+    /// The connection upgrade is performed in a background task. If that fails this callback
+    /// will be called.
+    ///
+    /// By default any errors will be silently ignored.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use axum::{
+    ///     extract::{WebSocketUpgrade},
+    ///     response::Response,
+    /// };
+    ///
+    /// async fn handler(ws: WebSocketUpgrade) -> Response {
+    ///     ws.on_failed_upgrade(|error| {
+    ///         report_error(error);
+    ///     })
+    ///     .on_upgrade(|socket| async { /* ... */ })
+    /// }
+    /// #
+    /// # fn report_error(_: axum::Error) {}
+    /// ```
+    pub fn on_failed_upgrade<C>(self, callback: C) -> WebSocketUpgrade<C>
+    where
+        C: OnFailedUpdgrade,
+    {
+        WebSocketUpgrade {
+            config: self.config,
+            protocol: self.protocol,
+            sec_websocket_key: self.sec_websocket_key,
+            on_upgrade: self.on_upgrade,
+            on_failed_upgrade: callback,
+            sec_websocket_protocol: self.sec_websocket_protocol,
+        }
+    }
+
+    /// Finalize upgrading the connection and call the provided callback with
+    /// the stream.
+    #[must_use = "to setup the WebSocket connection, this response must be returned"]
+    pub fn on_upgrade<C, Fut>(self, callback: C) -> Response
+    where
+        C: FnOnce(WebSocket) -> Fut + Send + 'static,
+        Fut: Future<Output = ()> + Send + 'static,
+        F: OnFailedUpdgrade,
+    {
+        let on_upgrade = self.on_upgrade;
+        let config = self.config;
+        let on_failed_upgrade = self.on_failed_upgrade;
+
+        let protocol = self.protocol.clone();
+
+        tokio::spawn(async move {
+            let upgraded = match on_upgrade.await {
+                Ok(upgraded) => upgraded,
+                Err(err) => {
+                    error!("Something wrong with on_upgrade: {:?}", err);
+                    on_failed_upgrade.call(Error::new(err));
+                    return;
+                }
+            };
+            let socket = WebSocketStream::from_raw_socket(
+                TokioAdapter::new(upgraded),
```

Comment:
This is the only other modified part where we add a TokioAdapter — in addition to the import of async-tungstenite at the top.

Reply:
I'd annotate this in the source file too.
So that if the upstream file changes, it is easier to update this file too.